File: C/21_Binary_Search/BOJ_07_12015_가장긴증가하는부분수열2.c
#include <stdio.h>
main() {
	int n, top = -1;
	scanf("%d", &n);
	int* arr = (int*)malloc(sizeof(int) * (n + 1));
	arr[++top] = 0;
	for (int i = 0; i < n; i++) {
		int value;
		scanf("%d", &value);
		if (value > arr[top]) {
			arr[++top] = value;
		}
		else {
			int left = 0, right = top, idx = 0;
			while (left <= right) {
				int mid = (right + left) / 2;
				if (arr[mid] >= value) {
					idx = mid;
					right = mid - 1;
				}
				else {
					left = mid + 1;
				}
			}
			arr[idx] = value;
		}
	}
	printf("%d", top);
}

File: JAVA/21_Binary_Search/BOJ_07_12015_가장긴증가하는부분수열2.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        ArrayList<Integer> list = new ArrayList<>();
        list.add(0);
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0; i < n; i++) {
            int value = Integer.parseInt(st.nextToken());
            if (value > list.get(list.size() - 1)) {
                list.add(value);
            } else {
                int left = 0, right = list.size() - 1, idx = 0;
                while (left <= right) {
                    int mid = (left + right) / 2;
                    if (list.get(mid) >= value) {
                        idx = mid;
                        right = mid - 1;
                    } else {
                        left = mid + 1;
                    }
                }
                list.set(idx, value);
            }
        }
        System.out.println(list.size() - 1);
    }
}